src/metabase/api/dashboard.clj

```diff
@@ -215,14 +216,97 @@
       hide-unreadable-cards
       add-query-average-durations))
 
+(defn- cards-to-copy
+  "Returns a map of which cards we need to copy and which are not to be copied. The `:copy` key is a map from id to
+  card. The `:discard` key is a vector of cards which were not copied due to permissions."
+  [ordered-cards]
+  (letfn [(split-cards [{:keys [card series] :as db-card}]
+            (cond
+              (nil? (:card_id db-card)) ;; text card
+              []
+
+              (mi/can-write? card)
+              (let [{writable true unwritable false} (group-by (comp boolean mi/can-write?)
+                                                               series)]
+                [(into [card] writable) unwritable])
+              ;; if you can't write the base, we don't have anywhere to put the series
+              :else
+              [[] (into [card] series)]))]
+    (reduce (fn [acc db-card]
+              (let [[retain discard] (split-cards db-card)]
+                (-> acc
+                    (update :copy merge (m/index-by :id retain))
+                    (update :discard concat discard))))
+            {:copy {}
+             :discard []}
+            ordered-cards)))
+
+(defn- duplicate-cards
+  "Takes a dashboard id, and duplicates the cards both on the dashboard's cards and dashcardseries. Returns a map of
+  {old-card-id duplicated-card} so that the new dashboard can adjust accordingly."
+  [dashboard dest-coll-id]
+  (let [same-collection? (= (:collection_id dashboard) dest-coll-id)
+        {:keys [copy discard]} (cards-to-copy (:ordered_cards dashboard))]
+    (reduce (fn [m [id card]]
+              (assoc m id
+                     (if (:dataset card)
+                       card
+                       (api.card/create-card!
+                        (cond-> (assoc card :collection_id dest-coll-id)
+                          same-collection?
+                          (update :name #(str % " -- " (tru "Duplicate"))))))))
+            {:uncopied discard}
+            copy)))
+
+(defn update-cards-for-copy
+  "Update ordered-cards in a dashboard for copying. If shallow copy, returns the cards. If deep copy, replaces ids with
+  id from the newly-copied cards. If there is no new id, it means user lacked curate permissions for the cards
+  collections and it is omitted. Dashboard-id is only needed for useful errors."
+  [dashboard-id ordered-cards deep? id->new-card]
+  (when (and deep? (nil? id->new-card))
```

Comment:
Is it better if we move this check to the API?
So that this function can be changed to `(defn reconcile-cards-for-deep-copy [ordered-cards id->new-card])` 
where its sole job is to replace ids from old to new IDs?

Reply:
I don't think so. The `id->new-card` is a map from the old id to the copied cards. At this point, if we're doing a shallow copy, there is no work to do. Otherwise we need to update the old ids to be the new ids. So this check is just if we are doing a deep copy, it is an error to not have the copied cards to substitute in. So I think this check here makes sense.